bson/_cbsonmodule.c

```diff
@@ -1743,6 +1743,42 @@ int write_dict(PyObject* self, buffer_t buffer,
         while (PyDict_Next(dict, &pos, &key, &value)) {
             if (!decode_and_write_pair(self, buffer, key, value,
                                     check_keys, options, top_level)) {
+                if (PyErr_Occurred()) {
+                    PyObject *etype, *evalue, *etrace;
+                    PyErr_Fetch(&etype, &evalue, &etrace);
+                    PyObject *InvalidDocument = _error("InvalidDocument");
+
+                    if (PyErr_GivenExceptionMatches(etype, InvalidDocument)) {
```

Comment:
Can you please reduce the nesting by using `if (InvalidDocument && PyErr_GivenExceptionMatches(etype, InvalidDocument))`?